psl-core/src/main/java/org/linqs/psl/util/RandUtils.java

```diff
@@ -176,4 +199,90 @@ public static synchronized void pairedShuffle(List... lists) {
             indexes[swapIndex] = indexes[i];
         }
     }
+
+    /**
+     * Sample from gridded points on the surface of a hypersphere with a radius 1.
+     * The grid is defined by
+     */
+    public static synchronized double[] sampleGriddedHypersphereSurface (int dimension, double gridStepSize) {
+        ensureRNG();
+
+        double[] vector = sampleHypersphereSurface(dimension, 1.0);
+
+        for (int i = 0; i < vector.length; i ++) {
+            vector[i] = Math.round(vector[i] / gridStepSize) * gridStepSize;
+        }
+
+        vector = MathUtils.toUnit(vector);
+
+        return vector;
+    }
+
+    /**
+     * Sample from the surface of a hypersphere with a radius of 1
+     */
+    public static synchronized double[] sampleUnitHypersphereSurface (int dimension) {
+        ensureRNG();
+
+        return sampleHypersphereSurface (dimension, 1.0);
+    }
+
+    /**
+     * Sample from the surface of a hypersphere with a certain radius
+     */
+    public static synchronized double[] sampleHypersphereSurface (int dimension, double radius) {
+        ensureRNG();
+
+        double[] vector = new double[dimension];
+
+        for (int i = 0; i < dimension; i++) {
+            // Returns Gaussian with mean = 0.0 and variance = 1.0
+            vector[i] = rng_gaussian.nextGaussian();
+        }
+
+        vector = MathUtils.toMagnitude(vector, radius);
+
+        return vector;
+    }
+
+    /**
+     * Sample from a dirichlet with the provided parameters
+     */
+    public static synchronized double[] sampleDirichlet (double[] alphas) {
+        ensureRNG();
+
+        double [] gammaSamples = new double[alphas.length];
+        double gammaSampleSum = 0;
+        double [] dirichletSample = new double[alphas.length];
+
+        for(int i = 0; i < alphas.length; i++){
+            gammaSamples[i] = sampleGamma(alphas[i],1);
+            gammaSampleSum = gammaSampleSum + gammaSamples[i];
+        }
+
+        for(int i = 0; i < alphas.length; i++){
+            dirichletSample[i] = gammaSamples[i] / gammaSampleSum ;
+        }
+
+        return dirichletSample;
+    }
+
+
+    /**
+     * Sample from a gamma with the provided parameters
+     */
+    public static synchronized double sampleGamma (double shape, double scale) {
+        ensureRNG();
+
+        if (gammaDist == null) {
```

Comment:
Ideally, you want utility functions like this to carry as little state as possible.
What happens when you have two different threads with different shapes iteratively calling on this function?

Instead, you should return a new object (seeded by the next long) that the caller can keep and generate new numbers from their distribution.